### PR TITLE
String to decimal conversion written using E/scientific notation

### LINKED ISSUE #5549: Add scientific notation decimal parsing in `parse_decimal`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
In Delta-RS we write min and max stats of a particular column, depending on the value of a decimal this is written in scientific notation. When we interact with a table we parse the logs to get a table state. As a result some users can't read tables that had written these stats.

See related issues:
-  https://github.com/delta-io/delta-rs/issues/2221
- https://github.com/delta-io/delta-rs/issues/2193

**Describe the solution you'd like**
Add support for scientific notation parsing.

**Describe alternatives you've considered**
None at the moment, our new table replay is deeply tied with the arrow json parser  :)

### PR DESCRIPTION
**Pull request by Nekit2217**
# Which issue does this PR close?

Closes #5549.

# Rationale for this change

There are errors when creating checkpoints in DeltaLake related to converting strings in E notation to a decimal field

# What changes are included in this PR?

Added string to decimal conversion written using E/scientific notation, no APIs were affected

# Are there any user-facing changes?

No

### CODE REVIEW

**Review by tustvold — COMMENTED** [review]
Thank you for this, could we perhaps get some tests of:

* Overflow behaviour
* Malformed numbers containing multiple `e` or multiple `+` / `-`

**Comment by Nekit2217** [comment]
Fixed a few bugs, didn't find any more bugs, if you have any comments, please let me know

**Comment by tustvold** [comment]
Sorry I am a bit swamped at the moment, I will review this first thing next week if nobody gets there first

File: arrow-cast/src/parse.rs [hunk 1]
```diff
@@ -671,6 +671,126 @@ impl Parser for Date64Type {
...
 }
 
+fn parse_e_notation<T: DecimalType>(
+    s: &str,
+    mut digits: u16,
+    mut fractionals: i16,
+    mut result: T::Native,
+    index: usize,
+    precision: u16,
+    scale: i16,
```
Comment (tustvold) [inline arrow-cast/src/parse.rs:681]: I wonder if this state could be encapsulated into a struct or similar, possibly something for a follow up PR

**Comment by tustvold** [comment]
So this does represent a non-trivial performance regression ~8%, but I suspect this is likely unavoidable and the regression isn't hugely concerning to me.

**Comment by ion-elgreco** [comment]
@Nekit2217 @tustvold thanks both!!

**Pull request merged**